inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -2,33 +2,1287 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-#include <cmath>
+#include "mkldnn_topk_node.h"
 
-#include <ngraph/op/topk.hpp>
+#include <mkldnn.hpp>
+
+#include <string>
+#include <vector>
+#include <set>
+#include <mkldnn_extension_utils.h>
+#include "emitters/jit_load_store_emitters.hpp"
 #include "ie_parallel.hpp"
-#include "mkldnn_topk_node.h"
-#include "utils/general_utils.h"
+#include <ngraph/op/topk.hpp>
+#include <algorithm>
+
+#include <cpu/x64/jit_generator.hpp>
+#include <cpu/x64/jit_uni_eltwise.hpp>
+#include "common/cpu_memcpy.h"
 
-#if defined(HAVE_SSE) || defined(HAVE_AVX2) || defined(HAVE_AVX512F)
-#include <immintrin.h>
-#endif
+#include <ngraph/opsets/opset1.hpp>
 
+using namespace mkldnn;
 using namespace MKLDNNPlugin;
 using namespace InferenceEngine;
+using namespace mkldnn::impl;
+using namespace mkldnn::impl::cpu::x64;
+using namespace mkldnn::impl::utils;
+using namespace Xbyak;
+
+#define GET_OFF(field) offsetof(jit_topk_call_args, field)
+
+#define vmm_tmp     Vmm(0)
+#define vmm_mask    Vmm(1)
+#define vmm_val(i)  Vmm(2 * (i) + 2)
+#define vmm_idx(i)  Vmm(2 * (i) + 3)
+#define vmm_val_l   Vmm(2)
+#define vmm_idx_l   Vmm(3)
+#define vmm_val_r   Vmm(4)
+#define vmm_idx_r   Vmm(5)
+
+#define xmm_tmp     Xmm(0)
+#define xmm_mask    Xmm(1)
+#define xmm_val(i)  Xmm(2 * (i) + 2)
+#define xmm_idx(i)  Xmm(2 * (i) + 3)
+#define xmm_val_l   Xmm(2)
+#define xmm_idx_l   Xmm(3)
+#define xmm_val_r   Xmm(4)
+#define xmm_idx_r   Xmm(5)
+
+#define xmm_val_p   Xmm(6)
+#define xmm_idx_p   Xmm(7)
+
+#define JMP_TO_LABEL(label)                  \
+    if (isa == avx512_common) {              \
+        kmovw(reg_tmp_32, k_mask);           \
+    } else {                                 \
+        uni_vmovmskps(reg_tmp_32, xmm_mask); \
+    }                                        \
+    and_(reg_tmp_32, 0x1);                   \
+    cmp(reg_tmp_32, 0x1);                    \
+    je(label, T_NEAR);
+
+static inline bool isFloatCompatible(memory::data_type type) {
+    return memory::data_type::f32 == type || memory::data_type::bf16 == type;
+}
+
+template <cpu_isa_t isa>
+struct jit_uni_topk_kernel_f32 : public jit_uni_topk_kernel, public jit_generator {
+    DECLARE_CPU_JIT_AUX_FUNCTIONS(jit_uni_topk_kernel_f32)
+
+    explicit jit_uni_topk_kernel_f32(jit_topk_config_params jcp)
+        : jit_uni_topk_kernel(jcp), jit_generator() {}
+
+    void create_ker() override {
+        jit_generator::create_kernel();
+        ker_ = (decltype(ker_))jit_ker();
+    }
+
+    void generate() override {
+        load_emitter.reset(new jit_load_emitter(this, isa, nullptr));
+        store_emitter.reset(new jit_store_emitter(this, isa, nullptr));
+        if (!mayiuse(avx512_core_bf16) && mayiuse(avx512_core))
+            emu_vcvtneps2bf16.reset(new jit_emu_vcvtneps2bf16(this, isa, nullptr));
```

Comment:
Maybe I missed smt but cannot find where bf16 emulator is used. Do we really need it?

Reply:
Fixed. Thanks Dmitry!